Update ngsi_arcgis_featuretable_sink.md (#2327)
telefonicaid/fiware-cygnus: doc/cygnus-ngsi/flume_extensions_catalogue/ngsi_arcgis_featuretable_sink.md

All this parameters, can be customized using Context Broker custom notifications (preferred) or Cygnus mapping capabilities.

Assuming that the feature table's url: `https://arcgis.com/{hash}/arcgis/rest/services/vehicles/cars` and feature table definition is:
```
Fields:

    objectid ( type: esriFieldTypeOID, alias: objectid, editable: false, nullable: false, defaultValue: null, modelName: objectid )
    licensePlate ( type: esriFieldTypeString, alias: licensePlate, editable: true, nullable: true, length: 255, defaultValue: null, modelName: licensePlate )
    speed ( type: esriFieldTypeDouble, alias: speed, editable: true, nullable: true, defaultValue: null, modelName: speed )
    oilLevel ( type: esriFieldTypeDouble, alias: oilLevel, editable: true, nullable: true, defaultValue: null, modelName: oilLevel )
```

Where `licensePlate` is the unique field to store the entityId value from the Context Broker.

Let's see both configuration options:

##### Using Context Broker custom notifications (preferred)

> ["CB custom notifications"](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#custom-notifications) is the preferred option because of its simplicity and not having to manage configuration (name-mappings) files in the server. 

Reviewer: ```suggestion
[CB custom notifications](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#custom-notifications) is the preferred option because of its simplicity and not having to manage configuration (name-mappings) files in the server. 
```

#### <a name="section1.2.1"></a>ArcGis databases naming conventions
Each entity type needs an url and an unique field to be persisted into the feature table.

> NGSIArcgisFeatureTableSink composes each table's url with entitie's `service` and `service path`, to provide multiple tables access. The final url is compounds by `cygnus-ngsi.sinks.arcgis-sink.arcgis_service_url`+`fiware-service`+`fiware-servicepath`. 

Reviewer: ```suggestion
NGSIArcgisFeatureTableSink composes each table's url with entitie's `service` and `service path`, to provide multiple tables access. The final url is composed of `cygnus-ngsi.sinks.arcgis-sink.arcgis_service_url`+`fiware-service`+`fiware-servicepath`. 
```

Unique field is provided to allow `NGSIArcgisFeatureTableSink` to update existant entities. NGSI `entity type` will be used as unique field name. This means that a feature named `type` in the Feature Table cannot be filled in by the sink. If Feature Table needs to persist the value of entity type it has to be in a feature different than `type`.

> All this parameters, can be customized using Cygnus mapping capabilities or Context Broker custom notifications.

Reviewer: ```suggestion
All this parameters, can be customized using Context Broker custom notifications (preferred) or Cygnus mapping capabilities (should be avoided except in justified cases).
```

Reviewer: In addition to this

* Rename "Using Cygnus" to "Using Cygnus name mappings"
* Change the order of subsections so first we have "Using Context Broker custom notifications" and next we have "Using Cygnus name mappings"

##### Using Cygnus

Let's see an example:	

###### Agent.conf file:

	agent.arcgis-sink.arcgis_service_url = https://arcgis.com/{hash}/arcgis/rest/services
###### Entity data:

	service = vehicles
	service-path = /4wheels
	entity-type = car
###### result

	Feature table url: https://arcgis.com/{hash}/arcgis/rest/services/vehicles/4wheels
	Table's unique field: car

[Top](#top)

> #### Using Context Broker custom notifications

Reviewer: Should be included in ToC above? It seem that four level sections (i.e. `####`) are included in ToC, if I have understood correctly.

Author: The title level was wrong, fixed in https://github.com/telefonicaid/fiware-cygnus/pull/2327/commits/deb80da960593b8e59aed2dbddfb0cc167f0edf4

> If a service path has different entitiy types, following the previous example `car` and `van`, they will need to persist in different Feature Layers, to avoid the limitation of one "fiware-servicepath" can only persist in a unique feature table, it can be used [CB custom notifications](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#custom-notifications) in which we can map a different value for `fiware-service` and `fiware-servicepath` headers parameters. 

Author: > Btw, why to use `car` and `van` instead of more canonical `Car` and `Van`? Maybe it's due to existing documentation? It could be changed to our canonical entity types?

The original documentation had the entity type in lower case, so I did not change it, but if we want to align it with the new format I can do it in this PR

###### Entity data:

	service = vehicles
	service-path = /4wheels
	entity-type = van

> For example, if the Feature table for "van" is `https://arcgis.com/{hash}/arcgis/rest/services/vehicles/van`,  the custom notif could be:

Reviewer: I'd suggest to include the full JSON of the subscription, including the `subject` object, so it can be seen how `"type": "van" is used in this case).

Reviewer: In addition, review the whole document to change:

* van -> Van
* car -> Car

and any other type so we use the current cannonical way of naming entity types (i.e. with first capital letter).

```
"httpCustom":{
  "url": "http://iot-cygnus:<source_arcgis_port>/notify",
  "headers": {
    "fiware-service": "vehicles",
    "fiware-servicepath": "/van"
  }
}
```

Using this functionality a different "fiware-servicepath" can be assigned to every entity type.

Moveover, to modify the unique field value of `type` attribute it can be used the [ngsi patching functionality](https://github.com/telefonicaid/fiware-orion/blob/master/doc/manuals/orion-api.md#ngsi-payload-patching):

```
"httpCustom":{
  "url": "http://iot-cygnus:<source_arcgis_port>/notify",
  "ngsi": {
    "type": "car"
  },
  "headers": {
    "fiware-service": "vehicles",
    "fiware-servicepath": "/van"
  }
}
```

###### result

	Feature table url: https://arcgis.com/{hash}/arcgis/rest/services/vehicles/van

> 	Table's unique field: car

Reviewer: This is the result of the second example? Why don't include the result of the first example?

It's a bit weird... maybe we should talk by skype.

    cygnus-ngsi.sinks.arcgis-sink.type = com.telefonica.iot.cygnus.sinks.NGSIArcgisFeatureTableSink
    cygnus-ngsi.sinks.arcgis-sink.channel = arcgis-channel
    cygnus-ngsi.sinks.arcgis-sink.enable_name_mappings = true

>     cygnus-ngsi.sinks.arcgis-sink.enable_name_mappings = false

Reviewer 2: Duplicated line in example
